pkg/tcpip/transport/tcp/endpoint.go

```diff
@@ -2564,7 +2570,13 @@ func (e *endpoint) shutdownLocked(flags tcpip.ShutdownFlags) tcpip.Error {
 			}
 			// Wake up any readers that maybe waiting for the stream to become
 			// readable.
-			e.waiterQueue.Notify(waiter.ReadableEvents)
+			events := waiter.ReadableEvents
+			if e.shutdownFlags&tcpip.ShutdownWrite == 0 {
+				// If ShutdownWrite is not set, write end won't close and
+				// we end up with a half-closed connection
+				events |= waiter.EventRdHUp
```

Comment:
I am not sure this is correct. Shutting down write should not result in a RdHUP as that is meant to be delivered when we receive a FIN from the peer to indicate no more data is expected from the peer.

Reply:
You probably want to notify this in rcv.handleSegments where we process the incoming FIN.

Reply:
I am not sure if we are on the same page here. The newly introduced code checks whether the `tcpip.ShutdownWrite` is **not** set in the `shutdownFlags` at a point where `shutdownFlags` most certainly have `tcpip.ShutdownRead` flag set (see line 2556). This would mean that the `shutdownLocked` function will close the endpoint for read but write will stay open resulting in a half-closed TCP connection with write opened and read closed, which should trigger POLLRDHUP by definition. Linux implementation of TCP uses a similar approach by [checking shutdown mask to set POLLRDHUP flag](https://github.com/torvalds/linux/blob/f348d70a324e15afc701a494f32ec468abb7d1eb/net/ipv4/tcp.c#L367-L368).

Nevertheless, it is possible that I am overthinking everything and if you would still like me to change this to a different approach, I would be happy to do it.